src/control/lib/hardware/pciutils/provider.go

```diff
@@ -0,0 +1,64 @@
+//
+// (C) Copyright 2024 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
+package pciutils
+
+import (
+	"context"
+
+	"github.com/daos-stack/daos/src/control/lib/hardware"
+	"github.com/pkg/errors"
+)
+
+type ctxKey string
+
+const (
+	AccessKey ctxKey = "pciutilsAccess"
+)
+
+type (
+	accessProvider interface {
+		PCIeCapsFromConfig(cfgBytes []byte, dev *hardware.PCIDevice) error
+		Cleanup()
+	}
+)
+
+func Init(parent context.Context) (context.Context, error) {
+	return context.WithValue(parent, AccessKey, &api{}), nil
+}
+
+func accessFromContext(ctx context.Context) (accessProvider, error) {
+	ap, ok := ctx.Value(AccessKey).(accessProvider)
+	if !ok {
+		return nil, errors.New("pciutils access not initialized")
+	}
+
+	return ap, nil
+}
+
+func Fini(ctx context.Context) {
+	ap, err := accessFromContext(ctx)
+	if err != nil {
+		return
+	}
+
+	ap.Cleanup()
+}
+
+func PCIeCapsFromConfig(ctx context.Context, cfgBytes []byte) (*hardware.PCIDevice, error) {
```

Comment:
Hmm, I don't mind a utility function necessarily, but I was thinking of something similar to the other lib/hardware providers, a struct with methods that could be mocked, rather than requiring higher level unit tests to know something about the (non-exported) `accessProvider` interface.

Reply:
we could revert @mjmac 's provider style of implementation but I don't feel that it's a particular problem. maybe I can visit that as a follow-up if it is deemed important.

Reply:
reworking...